File: legacy-typescript/src/runtime/store.ts
import { Type } from "../semantics/types.js";
import { RuntimeValue } from "./values.js";
import { defaultValue } from "./defaults.js";

let nextAddress = 1;

export interface CellLike {
  type: Type;
  get(): RuntimeValue;
  set(value: RuntimeValue): void;
  getRef?(): CellLike | undefined;
}

export class Cell implements CellLike {
  private ref?: Cell;
  readonly type: Type;
  private value: RuntimeValue;

  constructor(type: Type, value: RuntimeValue, ref?: Cell) {
    this.type = type;
    this.value = value;
    this.ref = ref;
  }

  get(): RuntimeValue {
    if (this.ref) return this.ref.get();
    return this.value;
  }

  set(value: RuntimeValue): void {
    if (this.ref) {
      this.ref.set(value);
      return;
    }
    this.value = value;
  }

  getRef(): Cell | undefined {
    return this.ref;
  }
}

export class Frame {
  private readonly slots = new Map<string, CellLike>();

  define(name: string, cell: CellLike): void {
    this.slots.set(name, cell);
  }

  lookup(name: string): CellLike | undefined {
    return this.slots.get(name);
  }
}

export class Store {
  readonly frames: Frame[] = [];
  readonly heap = new Map<number, CellLike>();
  readonly classHeap = new Map<number, { className: string; fields: Map<string, Cell> }>();
  readonly addresses = new Map<CellLike, number>();

  pushFrame(): Frame {
    const frame = new Frame();
    this.frames.push(frame);
    return frame;
  }

  popFrame(): void {
    this.frames.pop();
  }

  currentFrame(): Frame {
    if (this.frames.length === 0) {
      return this.pushFrame();
    }
    return this.frames[this.frames.length - 1];
  }

  allocPointerCell(type: Type): number {
    const cell = new Cell(type, defaultValue(type));
    const addr = nextAddress++;
    this.heap.set(addr, cell);
    return addr;
  }

  addrOf(cell: CellLike): number {
    const existing = this.addresses.get(cell);
    if (existing) return existing;
    const addr = nextAddress++;
    this.addresses.set(cell, addr);
    this.heap.set(addr, cell);
    return addr;
  }

  deref(addr: number): CellLike | undefined {
    return this.heap.get(addr);
  }

  allocClassObject(className: string, fields: Map<string, Cell>): number {
    const id = nextAddress++;
    this.classHeap.set(id, { className, fields });
    return id;
  }

  getClassObject(id: number): { className: string; fields: Map<string, Cell> } | undefined {
    return this.classHeap.get(id);
  }
}

File: csharp/Prescribe.Core/Runtime/Store.cs
using Prescribe.Core.Semantics;

namespace Prescribe.Core.Runtime;

public interface ICellLike
{
    TypeSymbol Type { get; }
    RuntimeValue Get();
    void Set(RuntimeValue value);
    ICellLike? GetRef();
}

public sealed class Cell : ICellLike
{
    private Cell? _ref;
    private RuntimeValue _value;

    public Cell(TypeSymbol type, RuntimeValue value, Cell? reference = null)
    {
        Type = type;
        _value = value;
        _ref = reference;
    }

    public TypeSymbol Type { get; }

    public RuntimeValue Get()
    {
        return _ref?.Get() ?? _value;
    }

    public void Set(RuntimeValue value)
    {
        if (_ref != null)
        {
            _ref.Set(value);
            return;
        }
        _value = value;
    }

    public ICellLike? GetRef() => _ref;
}

public sealed class Frame
{
    private readonly Dictionary<string, ICellLike> _slots = new(StringComparer.Ordinal);

    public void Define(string name, ICellLike cell) => _slots[name] = cell;

    public ICellLike? Lookup(string name) => _slots.TryGetValue(name, out var cell) ? cell : null;
}

public sealed class Store
{
    private int _nextAddress = 1;
    public List<Frame> Frames { get; } = new();
    public Dictionary<int, ICellLike> Heap { get; } = new();
    public Dictionary<int, (string ClassName, Dictionary<string, Cell> Fields)> ClassHeap { get; } = new();
    public Dictionary<ICellLike, int> Addresses { get; } = new();

    public Frame PushFrame()
    {
        var frame = new Frame();
        Frames.Add(frame);
        return frame;
    }

    public void PopFrame()
    {
        if (Frames.Count > 0) Frames.RemoveAt(Frames.Count - 1);
    }

    public Frame CurrentFrame()
    {
        if (Frames.Count == 0)
        {
            return PushFrame();
        }
        return Frames[^1];
    }

    public int AllocPointerCell(TypeSymbol type)
    {
        var cell = new Cell(type, Defaults.DefaultValue(type));
        var addr = _nextAddress++;
        Heap[addr] = cell;
        return addr;
    }

    public int AddrOf(ICellLike cell)
    {
        if (Addresses.TryGetValue(cell, out var existing)) return existing;
        var addr = _nextAddress++;
        Addresses[cell] = addr;
        Heap[addr] = cell;
        return addr;
    }

    public ICellLike? Deref(int addr) => Heap.TryGetValue(addr, out var cell) ? cell : null;

    public int AllocClassObject(string className, Dictionary<string, Cell> fields)
    {
        var id = _nextAddress++;
        ClassHeap[id] = (className, fields);
        return id;
    }

    public (string ClassName, Dictionary<string, Cell> Fields)? GetClassObject(int id)
    {
        return ClassHeap.TryGetValue(id, out var obj) ? obj : null;
    }
}
